components/canvas/canvas_paint_thread.rs

```diff
@@ -2,1137 +2,279 @@
-                 antialias: bool,
-                 canvas_id: CanvasId)
-                 -> IpcSender<CanvasMsg> {
+    pub fn start() -> IpcSender<CanvasMsg> {
         let (sender, receiver) = ipc::channel::<CanvasMsg>().unwrap();
-        let antialias = if antialias {
-            AntialiasMode::Default
-        } else {
-            AntialiasMode::None
-        };
         thread::Builder::new().name("CanvasThread".to_owned()).spawn(move || {
-            let mut painter = CanvasPaintThread::new(size, webrender_api_sender, antialias, canvas_id);
+            let mut canvas_paint_thread = CanvasPaintThread::new();
             loop {
-                let msg = receiver.recv();
-                match msg.unwrap() {
-                    CanvasMsg::Canvas2d(message, canvas_id) => {
-                        assert!(canvas_id == painter.canvas_id);
-                        match message {
-                            Canvas2dMsg::FillText(text, x, y, max_width) => painter.fill_text(text, x, y, max_width),
-                            Canvas2dMsg::FillRect(ref rect) => painter.fill_rect(rect),
-                            Canvas2dMsg::StrokeRect(ref rect) => painter.stroke_rect(rect),
-                            Canvas2dMsg::ClearRect(ref rect) => painter.clear_rect(rect),
-                            Canvas2dMsg::BeginPath => painter.begin_path(),
-                            Canvas2dMsg::ClosePath => painter.close_path(),
-                            Canvas2dMsg::Fill => painter.fill(),
-                            Canvas2dMsg::Stroke => painter.stroke(),
-                            Canvas2dMsg::Clip => painter.clip(),
-                            Canvas2dMsg::IsPointInPath(x, y, fill_rule, chan) => {
-                                painter.is_point_in_path(x, y, fill_rule, chan)
+                match receiver.recv() {
+                    Ok(msg) => {
+                        match msg {
+                            CanvasMsg::Canvas2d(message, canvas_id) => {
+                                canvas_paint_thread.process_canvas_2d_message(message, canvas_id);
+                            },
+                            CanvasMsg::Close(canvas_id) =>{
+                                canvas_paint_thread.canvases.remove(&canvas_id);
+                            },
+                            CanvasMsg::Create(creator, size, webrenderer_api_sender, antialias) => {
+                                let canvas_id = canvas_paint_thread.create_canvas(
+                                    size,
+                                    webrenderer_api_sender,
+                                    antialias
+                                );
+                                creator.send(canvas_id).unwrap();
+                            },
+                            CanvasMsg::Recreate(size, canvas_id) =>{
+                                canvas_paint_thread.canvases.get_mut(&canvas_id).unwrap().recreate(size);
```

Comment:
nit: `[&canvas_id]` instead of `get_mut(..).unwrap()` should be nicer and give a better error message.

Reply:
I wanted to do that, but I got error `error[E0596] : cannot borrow immutable indexed content as mutable`

Once you said that, I tried seeing if it would work now that I'm not pulling out a local reference, but it still did it.

I do agree that the way I did it is ugly, though.

Reply:
Ah, right... what about a `canvas_paint_thread.canvas(canvas_id)` function, which internally does:

```rust
self.canvases.get_mut(&canvas_id).expect("Bogus canvas id")
```

That should avoid some of the ugliness.

Reply:
Done. Thanks for the idea, that looks dramatically better. Pushed it alongside removing the outdated comment.